### DISCUSSION TITLE
Cardinality errors clean-up

### DISCUSSION THREAD

**Discussion by fantix** [geldata/gel#3350] [General]
Currently, we have:

 * `ResultCardinalityMismatchError`: raised by the compiler if the required cardinality mismatches the result, e.g. index expression returning more than one element.
 * `CardinalityViolationError`: raised by the server through `assert_single` and `assert_exists` checks, or the error message matches `^.*".*_cardinality_idx".*$`
 * `NoDataError`: raised by the client binding when no result is found in `query_required_single*` queries.
 * `InterfaceError`: raised by the client when:
   1. Parsing a query in `query_required_single*` calls but the result cardinality is not applicable
   2. Parsing a query in `query_single*` calls but the server raises `ResultCardinalityMismatchError`.

(client behavior is found in edgedb-python)

Elvis suggested - pls feel free to correct or elaborate - that we should have consistent cadinality-related error types, basically as "static" violations and runtime errors. In the meanwhile, `NoDataError` or "`TooManyResultsError`" from `query_single*` or `query_required_single*` means the result cardinality is unexpected.

**Comment by 1st1** [comment]
I think this all comes down to the 4th bullet: 

> i. Parsing a query in query_required_single* calls but the result cardinality is not applicable

For this we should raise ResultCardinalityMismatchError on the client side. Conceptually it's a compile-time error.

> ii. Parsing a query in query_single* calls but the server raises ResultCardinalityMismatchError.

For this we shouldn't warp the error in InterfaceError. Just propagate it as is.

Everything else can stay the same, imo we don't need to change the errors hierarchy.

**Reply by 1st1** [reply]
This is for both sync and async APIs.

We should also make sure `asyncio.wait_for()` works and doesn't wreck the client.
